flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/ResettableExternalBuffer.java

```diff
@@ -0,0 +1,699 @@
+		private final SimpleCollectingOutputView recordCollector;
+
+		// Can't use recordCollector.getCurrentOffset(), maybe the offset of recordCollector is
+		// disrupted by the attempt of record writing.
+		private long currentDataBufferOffset;
+		private int numBytesInLastBuffer;
+
+		private int recordCount;
+
+		private LinkedList<BufferIterator> iterators;
+
+		private InMemoryBuffer(AbstractRowSerializer serializer) {
+			this.segmentSize = memory.get(0).size();
+			this.freeMemory = new ArrayList<>(memory);
+			// serializer has states, so we must duplicate
+			this.serializer = (AbstractRowSerializer) serializer.duplicate();
+			this.recordBufferSegments = new ArrayList<>(memory.size());
+			this.recordCollector = new SimpleCollectingOutputView(this.recordBufferSegments,
+					new ListMemorySegmentSource(this.freeMemory), this.segmentSize);
+			this.recordCount = 0;
+			this.iterators = new LinkedList<>();
+		}
+
+		private void reset() {
+			this.currentDataBufferOffset = 0;
+			this.recordCount = 0;
+
+			// reset free and record segments.
+			this.freeMemory.addAll(this.recordBufferSegments);
+			this.recordBufferSegments.clear();
+
+			this.recordCollector.reset();
+
+			iterators.clear();
+		}
+
+		@Override
+		public void close() {
+			this.freeMemory.clear();
+			this.recordBufferSegments.clear();
+			iterators.clear();
+		}
+
+		public boolean write(BaseRow row) throws IOException {
+			try {
+				this.serializer.serializeToPages(row, this.recordCollector);
+				currentDataBufferOffset = this.recordCollector.getCurrentOffset();
+				numBytesInLastBuffer = this.recordCollector.getCurrentPositionInSegment();
+				recordCount++;
+
+				for (BufferIterator iterator : iterators) {
+					iterator.recordBuffer.updateLimitInLastSegment(numBytesInLastBuffer);
+				}
+
+				return true;
+			} catch (EOFException e) {
+				return false;
+			}
+		}
+
+		private ArrayList<MemorySegment> getRecordBufferSegments() {
+			return recordBufferSegments;
+		}
+
+		private long getCurrentDataBufferOffset() {
+			return currentDataBufferOffset;
+		}
+
+		private int getNumRecordBuffers() {
+			int result = (int) (currentDataBufferOffset / segmentSize);
+			long mod = currentDataBufferOffset % segmentSize;
+			if (mod != 0) {
+				result += 1;
+			}
+			return result;
+		}
+
+		private int getNumBytesInLastBuffer() {
+			return numBytesInLastBuffer;
+		}
+
+		private BufferIterator newIterator(int beginRow, long offset) {
+			checkArgument(offset >= 0, "`offset` can't be negative!");
+
+			RandomAccessInputView recordBuffer = new RandomAccessInputView(
+					this.recordBufferSegments, this.segmentSize, numBytesInLastBuffer);
+
+			BufferIterator iterator = new BufferIterator(beginRow, offset, recordBuffer);
+			iterators.add(iterator);
+			return iterator;
+		}
+
+		/**
+		 * Iterator of in memory buffer.
+		 */
+		public class BufferIterator implements MutableObjectIterator<BinaryRow>, Closeable {
```

Comment:
The name `BufferIterator` is the same with another one, super annoying when reading codes...